### DISCUSSION TITLE
Genetic strategy game coded in FXGL - Territory

### DISCUSSION THREAD

**Discussion by jmaler** [AlmasB/FXGL#1150] [Show and tell]
Hi everybody (especially Almas),

yesterday I published on Steam free demo of the game Territory. I coded it on the top of JavaFX/FXGL in last about 12 months (next to my common full time job which was quite a challenge :).

![store_top_image](https://user-images.githubusercontent.com/79770104/152559647-f8b6eaef-fcec-48e5-ba44-5d9cf6706e35.png) 

It is an isometric single player strategy game about genes and animal species. Animals are reproducing, eating and dying, you can mutate them and indirectly control them to achieve various goals like dominance in the ecosystem.
You can download it here: https://store.steampowered.com/app/1621980/Territory__animals_genetic_strategy/

I have to say that using FXGL was a great choice, it even forced me to learn Kotlin a bit :)). I was much more productive than I though, complete game was done in ~100MDs (I am more manager and don't code regularly), the biggest challenge (by far) was in the end the game design.
I still have to polish it a bit and add more content, but it is reasonably playable already. Full release is planned to March 3, 2022.

I will probably start to develop another game on the top of FXGL in 6 months or so, if someone is interested to join forces, let me know.

**Comment by AlmasB** [comment]
Hi @jmaler 

This looks really good! Thanks for the feedback, as well. I'm quite interested to know if there is anything, apart from game design, you found particularly difficult from the implementation perspective. Perhaps anything that can be extended / improved. Any feedback on details, whether positive or negative, would be appreciated.

Re: polishing, if there's one or two problems on your mind, I'd be happy to help think about a solution. Otherwise, I wish the best of luck with the release!

I'll post the game link on social media to help attract a few more people.

**Reply by jmaler** [reply]
Hi Almas, thanks for reply 👍
I don't think I found something particulary difficult, I remember that in the beginnings fxgl-samples helped me a lot, but after a month or two I more or less understood all parts of FXGL I used (but not 3D or physics which I didn't use).

I really like game loop and entity lifecycle, also scenes (which is more or less JavaFX), it was fine to have 5000 entites in game without any problems (then problems started, but it is much more than I need so I didn't do any investigation or optimisation).
I didn't use for example notificatioservice or dialogservice as they were very basic to what I needed and wrote my own independent versions.

There were things on the way that would make sense to improve, but I think that the most important is your strategy with the framework - whether to cover more features, or cover existing features better / more in deep (maybe even drop some).

Just some thoughts from the top of my head I was solving recently:
- FXGL.getAudioPlayer() could have removeAllMusic and/or Sound, it could also have get method returning sounds/music that "are there" or what is currently "playing" (now I have to manage it manually)
- missing method removeAllEventHandlers() on FXGL.getEventBus() (I needed to solve it when multple new games are started)
- to get list of existing saved games, it is required to load the content of saved files if I understand it well (as my saves are in MBs, it is not possible to load all of them just to find out their save date/label/game turn - so I am saving save file "metadata" to separate json)
- it is not possible to load generic asset via FXGL.getAssetLoader() - somethink loadFile() (I needed it couple of times for example to load .mp4 file for intro)
- not possible to change cutscene actors picture size (or it is but then it looks broken, I am not sure right now)

If you are interested, we can have a call and discuss this and other things more in detail - as I wrote, it is a lot about what direction you want to take.

**Reply by AlmasB** [reply]
Thanks for this, there is definitely useful stuff I can incorporate back into the project. If you are available within the next 2-3 weeks, feel free to email me e.g. [here](https://github.com/AlmasB/FXGL/blob/dev/CODE_OF_CONDUCT.md#enforcement) and we can schedule a Teams (or zoom) meeting.

**Comment by ImperaEtConquer** [comment]
Woooooooooow, looks extremely cool. I think that's coolest FXGL game so far? Really want to try it out. I hope it's not free, right? I'd buy that for sure!

**Reply by jmaler** [reply]
Hi, you can play demo for free (already available on Steam), the full game will be sold for something like 10-15$ (at least initially).
